flips/context-propagation.md

```diff
@@ -0,0 +1,208 @@
+# Context Propagation and Network Layer Refactor (Core Protocol)
+
+
+
+## Objective
+
+FLIP to separate the API through which components are started from the API through which
+they expose their current status. 
+
+Also refactor the networking layer and to make it agnostic to Flow IDs.
+
+## Current Implementation
+
+The [`ReadyDoneAware`](https://github.com/onflow/flow-go/blob/7763000ba5724bb03f522380e513b784b4597d46/module/common.go#L6) interface provides an interface through which components / modules can be started and stopped. Calling the `Ready` method should start the component and return a channel that will close when startup has completed, and `Done` should be the corresponding method to shut down the component.
+
+When the network layer receives a message, it will pass the message to the [`Engine`](https://github.com/onflow/flow-go/blob/7763000ba5724bb03f522380e513b784b4597d46/network/engine.go) registered on
+the corresponding channel by [calling the engine's `Process` method](https://github.com/onflow/flow-go/blob/d31fd63eb651ed9faf0f677e9934baef6c4d9792/network/p2p/network.go#L406), passing it the Flow ID of the message sender.
+
+### Potential problems 
+#### Non-idempotency of `ReadyDoneAware`
+The current `ReadyDoneAware` interface is misleading, as by the name one might expect that it is only used to check the state of a component. However, in almost all current implementations the `Ready` method is used to both start the component *and* check when it has started up, and similarly for the `Done` method. 
+
+This introduces issues of concurrency safety, as most implementations do not properly handle the case where the `Ready` or `Done` methods are called more than once. See [this example](https://github.com/onflow/flow-go/pull/1026).
```

Comment:
My point here is more about the semantics of `ReadyDoneAware`.

You are correct that the owner of a component should manager the lifecycle. The problem here is that we have not separated lifecycle management from state observation, because `ReadyDoneAware` currently implements both.

While only one owner should be able to start a component, multiple other components may want to wait for it to be ready. This is where it becomes a problem if calling `Ready` also starts the component each time.